awx/main/models/inventory.py

```diff
@@ -218,67 +221,87 @@ def get_group_children_map(self):
+        m = re.match(r"split(?P<offset>\d+)of(?P<step>\d+)", split_str)
+        if not m:
+            raise ParseError(_('Could not parse subset as split specification.'))
+        offset = int(m.group('offset'))
+        step = int(m.group('step'))
+        if offset > step:
+            raise ParseError(_('Split offset must be greater than total number of splits.'))
+        return (offset, step)
+
+    def get_script_data(self, hostvars=False, towervars=False, show_all=False, subset=None):
+        hosts_kw = dict()
+        if not show_all:
+            hosts_kw['enabled'] = True
+        fetch_fields = ['name', 'id', 'variables']
+        if towervars:
+            fetch_fields.append('enabled')
+        hosts = self.hosts.filter(**hosts_kw).order_by('name').only(*fetch_fields)
+        if subset:
+            if not isinstance(subset, six.string_types):
+                raise ParseError(_('Inventory subset argument must be a string.'))
+            if subset.startswith('split'):
+                offset, step = Inventory.parse_split_params(subset)
+                hosts = hosts[offset::step]
+            else:
+                raise ParseError(_('Subset does not use any supported syntax.'))
+
         data = dict()
+        all_group = data.setdefault('all', dict())
 
         if self.variables_dict:
-            all_group = data.setdefault('all', dict())
             all_group['vars'] = self.variables_dict
+
         if self.kind == 'smart':
-            if len(self.hosts.all()) == 0:
-                return {}
-            else:
-                all_group = data.setdefault('all', dict())
-                smart_hosts_qs = self.hosts.filter(**hosts_q).all()
-                smart_hosts = list(smart_hosts_qs.values_list('name', flat=True))
-                all_group['hosts'] = smart_hosts
+            all_group['hosts'] = [host.name for host in hosts]
         else:
-            # Add hosts without a group to the all group.
-            groupless_hosts_qs = self.hosts.filter(groups__isnull=True, **hosts_q)
-            groupless_hosts = list(groupless_hosts_qs.values_list('name', flat=True))
-            if groupless_hosts:
-                all_group = data.setdefault('all', dict())
-                all_group['hosts'] = groupless_hosts
+            # Keep track of hosts that are members of a group
+            grouped_hosts = set([])
 
             # Build in-memory mapping of groups and their hosts.
-            group_hosts_kw = dict(group__inventory_id=self.id, host__inventory_id=self.id)
-            if 'enabled' in hosts_q:
-                group_hosts_kw['host__enabled'] = hosts_q['enabled']
-            group_hosts_qs = Group.hosts.through.objects.filter(**group_hosts_kw)
-            group_hosts_qs = group_hosts_qs.values_list('group_id', 'host_id', 'host__name')
+            group_hosts_qs = Group.hosts.through.objects.filter(
+                group__inventory_id=self.id,
+                host__inventory_id=self.id
```

Comment:
here is the source of the bug @one-t mentioned. Group-host relationships are being fetched, even for hosts outside of the slice. That would be fine, if we maintained logic for cutting out those not used when populating the group host lists.

I thought about this logic a lot, because I have another issue in mind where groups aren't populated for smart inventories. The solution to that could be to perform this filter, only restricting to where the host is in the already-fetched host set. However, that would require enumerating the host ids in the query, which is prohibitively expensive. Same applies to prefetching. The same sort of dilemma faces us for the solution of this bug.